### PR TITLE
Raw modules bundles

### LINKED ISSUE #1828: Should Modules have raw bundles?
Currently, for each bundle defined in `Algebra.Bundles`, there is a corresponding raw bundle, consisting of the operations but without the laws. The same is not true for `Algebra.Module.Bundles`. Should they be added?

### PR DESCRIPTION
**Pull request by Taneb**
Closes #1828

Turns out that the raw module bundles only need to be indexed on `Set` rather than `Ring` or `RawRing`!

### CODE REVIEW

File: src/Algebra/Module/Bundles.agda [hunk 1]
```diff
@@ -80,8 +77,17 @@ record LeftSemimodule (semiring : Semiring r ℓr) m ℓm
...
+    ; _≉_ to _≉ᴹ_
     )
 
+  rawLeftSemimodule : RawLeftSemimodule Carrier m ℓm
+  rawLeftSemimodule = record
+    { _≈ᴹ_ = _≈ᴹ_
+    ; _+ᴹ_ = _+ᴹ_
+    ; _*ₗ_ = _*ₗ_
+    ; 0ᴹ = 0ᴹ
+    }
```
Comment (jamesmckinna) [inline src/Algebra/Module/Bundles.agda:89]: I would love it if such 'substructure' definitions were available without having to be written out in full, cf. #2255 ... eg here, the `+ᴹ-rawMonoid` structure is already available, only the additional `_*ₗ_` field needs defining, but Agda makes such `record` re-combination a bit crunchier than I would like:
```agda
rawLeftSemimodule = record { M ; _*ₗ_ = _*ₗ_ } where module M = RawMonoid +ᴹ-rawMonoid
```
is a bit of a mouthful, but still more 'informative' than the 'flattened' version?

File: src/Algebra/Module/Bundles.agda [hunk 2]
```diff
@@ -106,14 +112,23 @@ record LeftModule (ring : Ring r ℓr) m ℓm : Set (r ⊔ ℓr ⊔ suc (m ⊔ 
...
     using () renaming (group to +ᴹ-group; rawGroup to +ᴹ-rawGroup)
 
+  rawLeftModule : RawLeftModule Carrier m ℓm
+  rawLeftModule = record
+    { _≈ᴹ_ = _≈ᴹ_
+    ; _+ᴹ_ = _+ᴹ_
+    ; _*ₗ_ = _*ₗ_
+    ; 0ᴹ = 0ᴹ
+    ; -ᴹ_ = -ᴹ_
+    }
```
Comment (jamesmckinna) [inline src/Algebra/Module/Bundles.agda:130]: Ditto., et. seq.

File: src/Algebra/Module/Bundles.agda [hunk 3]
```diff
@@ -152,8 +163,17 @@ record RightSemimodule (semiring : Semiring r ℓr) m ℓm
...
     ; rawMonoid to +ᴹ-rawMonoid
+    ; _≉_ to _≉ᴹ_
     )
 
+  rawRightSemimodule : RawRightSemimodule Carrier m ℓm
+  rawRightSemimodule = record
+    { _≈ᴹ_ = _≈ᴹ_
+    ; _+ᴹ_ = _+ᴹ_
+    ; _*ᵣ_ = _*ᵣ_
+    ; 0ᴹ = 0ᴹ
```
Comment (jamesmckinna) [inline src/Algebra/Module/Bundles.agda:174]: Ditto.

File: src/Algebra/Module/Bundles.agda [hunk 4]
```diff
@@ -178,14 +198,23 @@ record RightModule (ring : Ring r ℓr) m ℓm : Set (r ⊔ ℓr ⊔ suc (m ⊔
...
   open AbelianGroup +ᴹ-abelianGroup public
     using () renaming (group to +ᴹ-group; rawGroup to +ᴹ-rawGroup)
 
+  rawRightModule : RawRightModule Carrier m ℓm
+  rawRightModule = record
+    { _≈ᴹ_ = _≈ᴹ_
+    ; _+ᴹ_ = _+ᴹ_
+    ; _*ᵣ_ = _*ᵣ_
+    ; 0ᴹ = 0ᴹ
+    ; -ᴹ_ = -ᴹ_
```
Comment (jamesmckinna) [inline src/Algebra/Module/Bundles.agda:215]: Ditto.

File: src/Algebra/Module/Bundles.agda [hunk 5]
```diff
@@ -220,7 +249,20 @@ record Bisemimodule (R-semiring : Semiring r ℓr) (S-semiring : Semiring s ℓs
...
     using ( +ᴹ-commutativeMonoid; +ᴹ-monoid; +ᴹ-semigroup; +ᴹ-magma; +ᴹ-rawMagma
-          ; +ᴹ-rawMonoid; _≉ᴹ_)
+          ; +ᴹ-rawMonoid; rawLeftSemimodule; _≉ᴹ_)
+
+  open RightSemimodule rightSemimodule public
+    using ( rawRightSemimodule )
+
+  rawBisemimodule : RawBisemimodule R.Carrier S.Carrier m ℓm
+  rawBisemimodule = record
+    { Carrierᴹ = Carrierᴹ
```
Comment (jamesmckinna) [inline src/Algebra/Module/Bundles.agda:259]: Is this field not redundant?

File: src/Algebra/Module/Bundles.agda [hunk 6]
```diff
@@ -220,7 +249,20 @@ record Bisemimodule (R-semiring : Semiring r ℓr) (S-semiring : Semiring s ℓs
...
+    ; _≈ᴹ_ = _≈ᴹ_
+    ; _+ᴹ_ = _+ᴹ_
+    ; _*ₗ_ = _*ₗ_
+    ; _*ᵣ_ = _*ᵣ_
+    ; 0ᴹ = 0ᴹ
```
Comment (jamesmckinna) [inline src/Algebra/Module/Bundles.agda:264]: See above... ditto.

File: src/Algebra/Module/Bundles.agda [hunk 7]
```diff
@@ -254,13 +296,26 @@ record Bimodule (R-ring : Ring r ℓr) (S-ring : Ring s ℓs) m ℓm
...
+    using (leftSemimodule; rightSemimodule; rawBisemimodule)
+
+  rawBimodule : RawBimodule R.Carrier S.Carrier m ℓm
+  rawBimodule = record
+    { _≈ᴹ_ = _≈ᴹ_
+    ; _+ᴹ_ = _+ᴹ_
+    ; _*ₗ_ = _*ₗ_
+    ; _*ᵣ_ = _*ᵣ_
+    ; 0ᴹ = 0ᴹ
+    ; -ᴹ_ = -ᴹ_
```
Comment (jamesmckinna) [inline src/Algebra/Module/Bundles.agda:317]: ...

File: src/Algebra/Module/Bundles/Raw.agda [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    _+ᴹ_ : Op₂ Carrierᴹ
+    _*ₗ_ : Opₗ R Carrierᴹ
+    0ᴹ : Carrierᴹ
+
+  +ᴹ-rawMonoid : RawMonoid m ℓm
+  +ᴹ-rawMonoid = record
+    { _≈_ = _≈ᴹ_
+    ; _∙_ = _+ᴹ_
+    ; ε = 0ᴹ
+    }
```
Comment (jamesmckinna) [inline src/Algebra/Module/Bundles/Raw.agda:43]: But this is unavoidable?

File: src/Algebra/Module/Bundles/Raw.agda [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    0ᴹ : Carrierᴹ
+    -ᴹ_ : Op₁ Carrierᴹ
+
+  rawLeftSemimodule : RawLeftSemimodule R m ℓm
+  rawLeftSemimodule = record
+    { _≈ᴹ_ = _≈ᴹ_
+    ; _+ᴹ_ = _+ᴹ_
+    ; _*ₗ_ = _*ₗ_
+    ; 0ᴹ = 0ᴹ
+    }
```
Comment (jamesmckinna) [inline src/Algebra/Module/Bundles/Raw.agda:69]: Etc.

**Review by jamesmckinna — APPROVED** [review]
This all looks great as-is, but if you can find a nice way to streamline the definition of the `Raw*` substructures, that would be great!

Similarly, long ago in advance of this issue being closed/a PR such as this being raised, I had left [a placeholder](https://github.com/agda/agda-stdlib/blob/bde655f8aa742ea82aa2c33e732353509d7dd397/src/Algebra/Module/Construct/Zero.agda#L40-L46) (commit https://github.com/agda/agda-stdlib/pull/1902/commits/5fbceb2151c566f7ad5f52ff7cda43f9c3f22543) in `Algebra.Module.Construct.Zero` for defining the `Raw*` substructures in-place for the `Zero` instance of `*Module`... 

... can this be tidied up now, or would you rather I did that after this is merged?

File: src/Algebra/Module/Bundles.agda (hunk 1 above)
Comment (Taneb) [reply src/Algebra/Module/Bundles.agda:89]: Would that work in this case? `RawMonoid` has different field names to `RawLeftSemimodule`...

File: src/Algebra/Module/Bundles.agda (hunk 5 above)
Comment (Taneb) [reply src/Algebra/Module/Bundles.agda:259]: Yes. My development process is write the type signature, write "name = ?", load file, press C-c C-r in the hole to get a record skeleton, fill in all the holes, delete the redundant fields. Sometimes I forget the last step.

File: src/Algebra/Module/Bundles.agda (hunk 1 above)
Comment (jamesmckinna) [reply src/Algebra/Module/Bundles.agda:89]: Ach! I think you're right. Maddening! 
Having to add the `renaming` directives would vitiate the whole suggestion. Grrr.

File: src/Algebra/Module/Bundles/Raw.agda (hunk 9 above)
Comment (jamesmckinna) [reply src/Algebra/Module/Bundles/Raw.agda:69]: Sigh..

File: src/Algebra/Module/Construct/Zero.agda [hunk 10]
```diff
@@ -38,11 +38,32 @@ module ℤero where
 open ℤero
```
Comment (jamesmckinna) [inline src/Algebra/Module/Construct/Zero.agda:38]: Is this directive now also redundant?

Comment (Taneb) [reply src/Algebra/Module/Construct/Zero.agda:38]: I'm not sure it wasn't redundant to begin with, but either way I'll remove it

File: src/Algebra/Module/Construct/TensorUnit.agda [hunk 11]
```diff
@@ -19,6 +19,62 @@ private
...
+rawBimodule {R = R} = record
+  { RawBisemimodule (rawBisemimodule {R = rawSemiring})
+  ; -ᴹ_ = -_
+  } where open RawRing R
+
+
+rawSemimodule : {R : RawSemiring c ℓ} → RawSemimodule _ c ℓ
+rawSemimodule {R = R} = rawBisemimodule {R = R}
+
+rawModule : {R : RawRing c ℓ} → RawModule _ c ℓ
```
Comment (jamesmckinna) [inline src/Algebra/Module/Construct/TensorUnit.agda:72]: Any particular reason (here, and throughout above) to omit `R` as the domain of coefficients in the return type? Could that simplify the subsequent definition body (probably not!?)

Comment (Taneb) [reply src/Algebra/Module/Construct/TensorUnit.agda:72]: It's because the parameter isn't `R` but `R.Carrier`, and that's a bother to extract

**Comment by jamesmckinna** [comment]
Looking even better now. Many thanks for all the additional commits!
(And: still a shame about all the repetition. A possible refactoring for another time?)

File: src/Algebra/Module/Construct/Zero.agda (hunk 10 above)
Comment (jamesmckinna) [reply src/Algebra/Module/Construct/Zero.agda]: (and I agree: it was redundant before! My mistake, then as now!)

File: src/Algebra/Module/Construct/TensorUnit.agda (hunk 11 above)
Comment (jamesmckinna) [reply src/Algebra/Module/Construct/TensorUnit.agda:72]: D'oh! Apologies once again...

File: CHANGELOG.md [hunk 12]
```diff
@@ -33,9 +33,49 @@ Deprecated names
...
 -----------
 
+* `Algebra.Module.Bundles.Raw`: raw bundles for module-like algebraic structures
+
 Additions to existing modules
 -----------------------------
 
+* In `Algebra.Module.Bundles`, raw bundles are re-exported and the bundles expose their raw counterparts.
+
+* In `Algebra.Module.Cosntruct.DirectProduct`:
```
Comment (jamesmckinna) [inline CHANGELOG.md:43]: typo: `Construct`

Comment (Taneb) [reply CHANGELOG.md:43]: Good catch!

**Comment by jamesmckinna** [comment]
:tada: If there are no further comments by CoB tomorrow, I'll hit `merge`...

**Review by JacquesCarette — APPROVED** [review]
This is both great and massively screams for automation. No human should have had to type all of that.

**Comment by jamesmckinna** [comment]
Sorry for not merging when I had said I would. 
*very* glad that @JacquesCarette was able to offer a second pair of eyes, though`!
(and looking forward to the 'smart(er)' automation PR downstream... ;-))

**Pull request merged**